packages/langium/src/grammar/langium-grammar-validator.ts

```diff
@@ -103,6 +113,20 @@ export class LangiumGrammarValidator {
         }
     }
 
+    checkGrammarForUnusedRules(grammar: Grammar, accept: ValidationAcceptor): void {
+        let entryRule = true;
+        for (const rule of grammar.rules) {
+            if (isParserRule(rule) && entryRule) {
+                entryRule = false;
+                continue;
+            }
+            const references = this.references.findReferences(rule);
```

Comment:
This criterion is not sufficient. We want to mark rules as unreachable if they are referenced only by other unreachable rules.

Algorithmic approach: start a DFS from the entry rule and mark all visited rules. Then find all unmarked rules and give them a warning.